[RequirementMachine] Diagnose conflicting requirements.

hborla opened the pull request:
This change implements conflict diagnostics for pairs of rules that are marked as conflicting by the `PropertyMap`, including:
* Two concrete same-type requirements that are mismatched
* Two different superclass requirements
* A concrete same-type requirement that conflicts with a conformance requirement
* A concrete same-type requirement that conflicts with a superclass requirement
* A concrete same-type requirement that conflicts with a layout requirement

Resolves: rdar://88135471

hborla committed:
[RequirementMachine] Add a RequirementError for diagnosing a conflicting
same-type requirement.

hborla committed:
[RequirementMachine] Compute and diagnose conflicting rules in the minimal
rewrite system of a generic/requirement signature.

hborla committed:
[RequirementMachine] Add a helper method to RequirementMachine to compute
all requirement diagnostics from the minimal rewrite system.

hborla committed:
[RequirementMachine] Mention the type parameter that is the subject of
two conflicting requirements in diagnostics where possible.

hborla committed:
[RequirementMachine] Consolidate the various RequirementError::Kind cases
that represent conflict diagnostics.

slavapestov commented on lib/AST/RequirementMachine/RewriteSystem.cpp:
```diff
@@ -693,6 +694,105 @@ void RewriteSystem::computeRedundantRequirementDiagnostics(
+
+    auto firstProperty = firstRule->isPropertyRule();
+    auto secondProperty = secondRule->isPropertyRule();
+    if (!firstProperty || !secondProperty)
```
Can you change these to asserts since we only expect property rules here?

slavapestov commented on lib/AST/RequirementMachine/RewriteSystem.cpp:
```diff
@@ -693,6 +694,105 @@ void RewriteSystem::computeRedundantRequirementDiagnostics(
+    TypeArrayView<GenericTypeParamType> genericParams) {
+  for (auto pair : ConflictingRules) {
+    auto *firstRule = &getRule(pair.first);
+    auto *secondRule = &getRule(pair.second);
```
Why do these need to be pointers? Also should be const IMO

slavapestov commented on lib/AST/RequirementMachine/RewriteSystem.cpp:
```diff
@@ -693,6 +694,105 @@ void RewriteSystem::computeRedundantRequirementDiagnostics(
+      case Symbol::Kind::AssociatedType:
+      case Symbol::Kind::GenericParam:
+      case Symbol::Kind::Name:
+        return;
```
Assert if we hit one of these since we don't expect them to occur

slavapestov commented on test/Generics/requirement_machine_diagnostics.swift:
```diff
@@ -197,4 +197,109 @@ func inferred6<T : P11>(_: T) where T.Y : Hashable, T.Z == Set<T.X>, T.X == T.Y
+  var bar: Bar { get }
+}
+
+func sameTypeConflicts() {
```
I like using local types/functions nested inside of another function to group tests!

Can you add a couple of tests for conflicts inside protocol `where` clauses as well just in case there's something wrong with the getTypeForTerm() usage?

hborla committed:
[RequirementMachine] When computing conflict diagnostics, only compute the
type for the longer subject term between the two conflicting rules.

hborla committed:
[RequirementMachine] Simplify computing and emitting conflict diagnostics.

Instead of computing different combinations of conflicting requirement kinds,
emit the same error message for all of them.

slavapestov commented on lib/AST/RequirementMachine/RequirementLowering.cpp:
```diff
@@ -605,6 +590,37 @@ bool swift::rewriting::diagnoseRequirementErrors(
+        llvm::raw_string_ostream OS(requirements);
+        OS << "'";
+        requirement.print(OS, options);
+        OS << "' and '";
```
Perhaps “and” should be part of the diagnostic string itself?

slavapestov commented on lib/AST/RequirementMachine/Rule.cpp:
```diff
@@ -43,6 +43,33 @@ Optional<Symbol> Rule::isPropertyRule() const {
+  switch (property->getKind()) {
+  case Symbol::Kind::ConcreteType:
+    return Requirement(RequirementKind::SameType, subject,
+                       property->getConcreteType());
```
If the concrete type is generic, getConcreteType() returns a type where type parameters have been replaced with fresh generic parameters. Try something like `T == G<U.T>` vs `T == Int`. You need to use PropertyMap::getTypeFromSubstitutionSchema(), passing in symbol->getConcreteType() and symbol->getSubstitutions().

slavapestov replied:
The 'prefix' parameter to getTypeFromSubstitutionSchema() is the prefix of the subject type whose length is the difference in length between the two terms. Suppose you have two conflicting property rules:

- `[P:X].[concrete: G1<[P:Y]>] => [P:X]` from `protocol P { associatedtype X where X == G1<Y>; associatedtype Y }`
- `[Q:X].[P:X].[concrete: G2<[Q:Z]>] => [Q:X].[P:X]` from `protocol Q { associatedtype X : P where X.X == G2<Z>; associatedtype Z }`

The subject type here is the longer of the two right hand sides, `[Q:X].[P:X]`. When converting the concrete type symbol from the first rule `[concrete: G<[P:Y]>]` into a Type, we have to prepend the prefix `[Q:X]` to the substitution term `[P:Y]` to get `[concrete: G<[Q:X].[P:Y]>]`, since the 'adjusted' concrete type requirement is 'X.X == G1<X.Y>` not `X.X == G1<Y>`.

slavapestov commented on lib/AST/RequirementMachine/Rule.cpp:
```diff
@@ -43,6 +43,33 @@ Optional<Symbol> Rule::isPropertyRule() const {
   return property;
 }
 
+Optional<Requirement> Rule::getPropertyRequirement(Type subject) const {
```
This looks a lot like the isPropertyRule() part of the getRequirementFromRule lambda in RequirementBuilder::addRequirement()… I wonder if we can merge them together somehow. Or maybe this can just be a static method in this file, since it’s only used in one place.

slavapestov commented on lib/AST/RequirementMachine/RewriteSystem.cpp:
```diff
@@ -15,6 +15,7 @@
 #include "llvm/Support/raw_ostream.h"
 #include <algorithm>
 #include <vector>
+#include "PropertyMap.h"
```
This means that RewriteSystem.cpp and PropertyMap.cpp include each other’s header, which is a bit funny. Perhaps the diagnostics methods could go in a new Diagnostics.cpp, along with the bits from RequirementLowering.cpp and the rule::getPropertyRequirement() method

slavapestov reviewed (commented):
This looks great, only a few remaining comments

slavapestov commented on lib/AST/RequirementMachine/Diagnostics.h:
```diff
@@ -41,11 +42,20 @@ struct RequirementError {
 
+  /// A requirement that conflicts with \c requirement. Both
+  /// requirements will have the same subject type.
+  Optional<Requirement> conflictingRequirement;
```
Is it worth asserting they have the same subject type in the constructor?

slavapestov commented on test/Generics/requirement_machine_diagnostics.swift:
```diff
@@ -197,4 +197,112 @@ func inferred6<T : P11>(_: T) where T.Y : Hashable, T.Z == Set<T.X>, T.X == T.Y
+}
+
+func basicConflict<T: ProtoAlias1 & ProtoAlias2>(_:T) where T.A1 == T.A2 {}
+// expected-error@-1{{no type for 'T.A1' can satisfy both 'T.A1 == Int' and 'T.A1 == String'}}
```
Oh, "no type for T" is a great way to say "type parameter T" without saying "type parameter". Nice!

hborla committed:
[RequirementMachine] Record the sugared 'Self' type parameter when
initializing a RequirementMachine for a written protocol signature.

These generic parameters are used for re-sugaring when computing a
type from a term, so if the recorded 'Self' parameter is canonicalized,
it shows up in diagnostics as 'tau_0_0'.

hborla committed:
[NFC][RequirementMachine] Stylistic changes for computing confict
diagnostics.

Move computing requirements from rules into a static function, make
a few variables const, etc.

hborla committed:
[RequirementMachine] Use PropertyMap::getTypeFromSubstitutionSchema when
computing a concrete same-type or superclass for conflict diagnostics.

Otherwise, diagnostics will show fresh type parameters when the concrete
type is generic.

hborla committed:
[RequirementMachine] When computing requirements for conflict diagnostics,
adjust the concrete type symbol for the suffix rule by applying the prefix
from the subject rule.

hborla committed:
[RequirementMachine] Omit duplicate conflict diagnostics.

hborla commented:
@swift-ci please smoke test

hborla commented:
@swift-ci please test Windows

hborla commented:
The Windows failure is unrelated

>C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\llvm-project\lldb\source\Plugins\Platform\Android\PlatformAndroid.cpp(146): error C2512: 'lldb_private::platform_android::PlatformAndroidRemoteGDBServer::PlatformAndroidRemoteGDBServer': no appropriate default constructor available
